components/self_hosted_purchase_modal/index.tsx

```diff
@@ -1,56 +1,520 @@
+    SelfHostedSignupCustomerResponse,
+} from '@mattermost/types/hosted_customer';
+import {UserProfile} from '@mattermost/types/users';
 import {ValueOf} from '@mattermost/types/utilities';
-import {getSelfHostedSignupProgress} from 'mattermost-redux/selectors/entities/cloud';
+import {getTheme} from 'mattermost-redux/selectors/entities/preferences';
+import {getConfig} from 'mattermost-redux/selectors/entities/general';
+import {getAdminAnalytics} from 'mattermost-redux/selectors/entities/admin';
+import {getSelfHostedProducts, getSelfHostedSignupProgress} from 'mattermost-redux/selectors/entities/hosted_customer';
+import {getCurrentUser} from 'mattermost-redux/selectors/entities/users';
+import {Client4} from 'mattermost-redux/client';
+import {HostedCustomerTypes} from 'mattermost-redux/action_types';
 import {DispatchFunc} from 'mattermost-redux/types/actions';
 
 import {trackEvent, pageVisited} from 'actions/telemetry_actions';
 import {closeModal} from 'actions/views/modals';
+import {confirmSelfHostedSignup} from 'actions/hosted_customer';
 
 import {GlobalState} from 'types/store';
 
 import {isModalOpen} from 'selectors/views/modals';
 
-import {STRIPE_CSS_SRC} from 'components/payment_form/stripe';
+import {COUNTRIES} from 'utils/countries';
 
 import {
     ModalIdentifiers,
+    StatTypes,
     TELEMETRY_CATEGORIES,
 } from 'utils/constants';
 
+import CardInput, {CardInputType} from 'components/payment_form/card_input';
+import StateSelector from 'components/payment_form/state_selector';
+import DropdownInput from 'components/dropdown_input';
+import BackgroundSvg from 'components/common/svg_images_components/background_svg';
+import UpgradeSvg from 'components/common/svg_images_components/upgrade_svg';
+
+import Input from 'components/widgets/inputs/input/input';
+
 import FullScreenModal from 'components/widgets/modals/full_screen_modal';
 import RootPortal from 'components/root_portal';
 import useLoadStripe from 'components/common/hooks/useLoadStripe';
+import useFetchStandardAnalytics from 'components/common/hooks/useFetchStandardAnalytics';
+
+import StripeProvider from './stripe_provider';
+import SelfHostedCard from './self_hosted_card';
+import SuccessPage from './success_page';
+import ContactSalesLink from './contact_sales_link';
+import Submitting, {convertProgressToBar} from './submitting';
+import ErrorPage from './error';
+import {Seats, errorInvalidNumber} from './seats_calculator';
+import Terms from './terms';
+import useNoEscape from './useNoEscape';
+
+import './self_hosted_purchase_modal.scss';
+
+interface State {
+    address: string;
+    address2: string;
+    city: string;
+    state: string;
+    country: string;
+    postalCode: string;
+    cardName: string;
+    organization: string;
+    agreedTerms: boolean;
+    cardFilled: boolean;
+    seats: Seats;
+    submitting: boolean;
+    succeeded: boolean;
+    progressBar: number;
+    error: string;
+}
+
+interface UpdateAddress {
```

Comment:
This suggestion can be integrated or not. It has a tradeoff that can't be ignored . We can take advantage of TS and reduce this line to 177. The suggestion is the following:
```typescript
type UpdateAction<K extends string, V> = {
  type: `update_${K}`;
  data: V;
}

type UnionUpdateActions<T> = {
  [Key in Extract<keyof T, string>]: UpdateAction<Key, T[Key]>
}[Extract<keyof T, string>]
```
I've created a [playground to test it](https://www.typescriptlang.org/play?#code/C4TwDgpgBAyhCGwDOUC8UnAE4EsB2A5gNwBQJ+wEWAZvAMbQzCLQDeJUnU8AJj1hCRIAXBmz5iHLr36CkAJlGZchUlyh0coJeNVTOmFjpWT1dAPYBXPNhDGJarmHOGANgGFzPCPb1n4WDwAcvAAtj5iJo6c5lgE8Hg4AF6IOOZ4vqbSBAIQPAAqVKEiUABG5uauCHjRGgE8AGI4rlU8ouWV1bVICMiicIhI3ZaloVrAEu0VVQnDdAx5eVOds-pQYFjmOXIAQgGieJahpVS1VJtYmaQAvmSgkFAAqmA8LACCdBPpADwA0lAQAAelDwPBQygkABooAA1AB8aCg7E49wiAANLC8WAB9AAkrF+1zRjlezFEMJud3A0EeiXSz1JEA+XzwSG++QR6GRUAA2r8ICAoPgoABRYFYejAb4AawF5moUHy0IhhDhAF1RAz3p80ng-gLofk+QK1XCSNceWLsJKZXKFUrIhJ1WQLKzgNxNXS8FrKMzdWymCxOUipKjRAByTGM7E9QbhyFkW5AA)

On the one hand, we have your approach, It’s better than my suggestion because it's easier to follow and maintain than the suggestion. On the other hand, the suggestion has a better DX as if you made a change, and TS automatically does the job for us

For me it's perfect both approach, as this case is DX vs Maintenance. If we are going to modify it frequently, I will suggest don't add it.

Reply:
I chose to implement your suggestion, and also added some helpers to derive a state setter.